Docs: New App plugin tutorial (#914)
grafana/plugin-tools: docusaurus/docs/tutorials/build-an-app-plugin.md

---
id: build-an-app-plugin
title: Build an app plugin
sidebar_position: 10
description: Learn how to create an app plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - app
---

import CreatePlugin from '@shared/create-plugin-frontend.md';
import PluginAnatomy from '@shared/plugin-anatomy.md';

## Introduction

> In this tutorial you will learn how to create _app plugins_. App plugins are Grafana plugins that allow you to bundle data sources and panel plugins within a single package. They also enable you to create custom pages within Grafana, providing a dedicated space for documentation, sign-up forms, or integration with other services via HTTP.

Reviewer: could reference scenes and tabs for those pages here

Author: @sympatheticmoose I am not sure what you mean here. What tabs do you refer to?

Reviewer: https://grafana.com/developers/scenes/scene-app-tabs

> App plugins are displayed in the navigation menu and offer the flexibility to define [UI Extensions](../ui-extensions/).

Reviewer: Are they always?

Reviewer 2: It would be useful for some further explanation here. At the start we say app plugins allow you to bundle data sources and panel plugins. If that's my intended use case, I probably don't want my app plugin to be displayed in the navigation menu; is this optional?

Author: I changed the wording to note it is optional

### Prerequisites

- Grafana v9.0 or later
- [LTS](https://nodejs.dev/en/about/releases/) version of Node.js

## Create a new app plugin

<CreatePlugin pluginType="app" />

:::note

> If you choose to have a backend for your app plugin, you must run `mage -v` before starting Grafana with Docker.

Reviewer 2: ```suggestion
If you choose to have a backend for your app plugin, you must ensure to build the binary by running `mage -v` before starting Grafana with Docker.
```

Additional context for why a developer has to run `mage -v`.

:::

## Anatomy of a plugin

<PluginAnatomy />

## Development workflow

Next, you'll learn the basic workflow of making a change to your app, building it, and reloading Grafana to reflect the changes you made.

The first step is to add your panel to a dashboard:

> 1. Open Grafana in your browser.

Reviewer 2: Missing a step here to run the Grafana development environment with `docker compose up --build`.

Author: Good point!. There's no need to pass `--build`. docker compose will build the image the first time.

1. Go to **Apps** -> **Your App Name**.

Now that you can view your app root page (page one in the example), try making a change to the app plugin:

1. In `PageOne.tsx`, change some of the page text content:

   ```tsx title="src/pages/PageOne.tsx"
   <PluginPage>New page content</PluginPage>
   ```

1. Save the file.

> 1. Reload Grafana. Your changes should appear in the browser.

Reviewer 2: > Reload Grafana
Does this mean refresh the browser, or `docker compose down && docker compose up --build`?

Author: Good point!

1. Restart your Grafana instance.

:::note

After saving the `plugin.json` file, you need to restart your Grafana instance to see the new page in the navigation menu.

:::

The new page appears in the navigation menu. You can now edit the React router in `src/components/App/App.tsx` and point a custom component to it.

1. Create a new file called `src/pages/NewPage.tsx` and add the following code:

   ```tsx title="src/pages/NewPage.tsx"
   import React from 'react';
   import { PluginPage } from '@grafana/runtime';

   export function NewPage() {
     return <PluginPage>New Page</PluginPage>;
   }
   ```

1. Modify the routes in `src/components/App/App.tsx` to recognize the new page:

   ```tsx title="src/components/App/App.tsx"
   {
     /* .... */
   }
   <Route path="new-page" element={<NewPage />} />;
   ```

1. Save the file.
1. Reload Grafana to see the new page in place.

You don't need to register all your pages inside `includes` in your `plugin.json`. Register only pages that you want to add to the navigation menu.

:::note

You can only have one level of pages in the navigation menu. Sub-menu items are not supported.

:::

## Configuration page

> You can add a configuration page to your App plugin where your users can store any settings your plugin might need. Your plugin should already have an example configuration page with its source code located in `src/components/AppConfig/AppConfig.tsx`.

Reviewer 2: ```suggestion
You can add a configuration page to your App plugin where users can configure any settings your plugin might need. Your plugin should already have an example configuration page with its source code located in `src/components/AppConfig/AppConfig.tsx`.
```

### Saving user settings

> To store user settings, send a POST request to `/api/plugins/%PLUGIN_ID%/settings` with the `jsonData` and `secureJsonData` as data.

Reviewer 2: ```suggestion
To save user settings, send a POST request to `/api/plugins/%PLUGIN_ID%/settings` with the `jsonData` and `secureJsonData` as data.
```

Nit picking here, but reduces confusion by sticking to consistent terminology.

```ts
export const updatePluginSettings = async (pluginId: string, data: Partial<PluginMeta>) => {
  const response = await getBackendSrv().fetch({
    url: `/api/plugins/${pluginId}/settings`,
    method: 'POST',
    data, // data: { jsonData: { ... }, secureJsonData: { ... } }
  });

  return lastValueFrom(response);
};
```

### Retrieve user settings

The user settings are part of the plugin `meta`. You can retrieve them inside a React component by using the `usePluginContext` hook. For example:

```tsx
import React from 'react';
import usePluginContext from '@grafana/data';

function MyComponent() {
  const context = usePluginContext();
  // user settings
  const jsonData = context.meta.jsonData;
}
```

## Fetch data from frontend code using the data proxy

> If you want to fetch data from your app frontend code without [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) issues or using authenticated requests, then you want to use the [data proxy](../create-a-plugin/extend-a-plugin/fetch-data-from-frontend.md).

Author: I agree a code example would be nice but that makes keeping the documentation up to date much harder with the same sample scattered around different pages.  I'll add a better explanation but keep the link to the page fully explaining data fetching.

> ## Bundle plugins inside your app
> 
> You can bundle data sources and panel plugins inside your app plugins. See [Work with nested plugins](../create-a-plugin/extend-a-plugin/nested-plugins).

Reviewer 2: ```suggestion
## Adding nested plugins inside your app

You can nest data sources and panel plugins inside your app plugin to provide a complete user experience. See [Work with nested plugins](../create-a-plugin/extend-a-plugin/nested-plugins).
```

May be worth using the term "nested" here instead of introducing the term "bundle" for consistency.

## Include external plugins

If you want to let users know that your app requires an existing plugin, you can add it as a dependency in `plugin.json`. Note that they'll still need to install it themselves.

```json title="src/plugin.json"
"dependencies": {
  "plugins": [
    {
      "type": "panel",
      "name": "Worldmap Panel",

>       "id": "grafana-worldmap-panel",

Reviewer: Please pick a different, non deprecated plugin 😄

Reviewer 2: My understanding was that plugins listed inside the dependencies were automatically installed by Grafana when you installed the "parent" plugin. Is this not the case?

Reviewer: that is the case - good catch Tom

      "version": "^0.3.2"
    }
  ]
}

> ```

Reviewer 2: Adding a conclusion and next steps section would round off this tutorial.

e.g.
- Signing your app plugin
- Publishing your app plugin

Author: nice idea!

---
id: build-an-app-plugin
title: Build an app plugin
sidebar_position: 10
description: Learn how to create an app plugin.
keywords:
  - grafana
  - plugins
  - plugin
  - app
---

import CreatePlugin from '@shared/create-plugin-frontend.md';
import PluginAnatomy from '@shared/plugin-anatomy.md';

## Introduction

> In this tutorial you will learn how to create _app plugins_. App plugins are Grafana plugins that allow you to bundle data sources and panel plugins within a single package. They also enable you to create custom pages within Grafana, providing a dedicated space for documentation, sign-up forms, integration with other services via HTTP or create [Scenes apps](https://grafana.com/developers/scenes/).

Reviewer 3: ```suggestion
In this tutorial you will learn how to create an _app plugin_. App plugins are Grafana plugins that allow you to bundle data sources and panel plugins within a single package. They also enable you to create custom pages within Grafana, providing a dedicated space for documentation, sign-up forms, integration with other services via HTTP. You can also use app plugins to create [Scenes apps](https://grafana.com/developers/scenes/).
```

> App plugins can be displayed in the [navigation menu](#add-a-page-in-the-navigation-menu) and offer the flexibility to define [UI Extensions](../ui-extensions/).

Reviewer 3: ```suggestion
App plugins can be displayed in the [navigation menu](#add-a-page-in-the-navigation-menu) and offer the flexibility to define [UI extensions](../ui-extensions/).
```

### Prerequisites

- Grafana v9.0 or later
- [LTS](https://nodejs.dev/en/about/releases/) version of Node.js

## Create a new app plugin

<CreatePlugin pluginType="app" />

:::note

If you choose to have a backend for your app plugin, you must build the backend plugin binary by running `mage -v` before starting Grafana with Docker.

:::

## Anatomy of a plugin

<PluginAnatomy />

## Development workflow

Next, you'll learn the basic workflow of making a change to your app, building it, and reloading Grafana to reflect the changes you made.

The first step is to see your scaffolded plugin in action.

> 1. Start your Grafana instance with `docker compose up`

Reviewer 3: ```suggestion
1. Start your Grafana instance with `docker compose up`.
```

1. Open Grafana in your browser. Go to [http://localhost:3000](http://localhost:3000).
1. Go to **Apps** -> **Your App Name**.

Now that you can view your app root page (page one in the example), try making a change to the app plugin:

1. In `PageOne.tsx`, change some of the page text content:

   ```tsx title="src/pages/PageOne.tsx"
   <PluginPage>New page content</PluginPage>
   ```

1. Save the file.
1. Reload Grafana in your browser. Your changes should appear.

## Add a page in the navigation menu

To add pages to the navigation menu under your app menu item, modify the `plugin.json` file's [`includes` section](../reference/metadata.md#includes).

> When you scaffold your plugin, the `create-plugin` tool adds some example pages to the navigation menu. Each of them follow a path such as `/a/%PLUGIN_ID%/PAGE_NAME`. Any requests sent to /a/%PLUGIN_ID%, for example, `/a/myorgid-simple-app/`, are routed to the root page of the app plugin. The root page is a React component that returns the content for a given route.

Reviewer 3: ```suggestion
When you scaffold your plugin, the `create-plugin` tool adds some example pages to the navigation menu. Each example page follows a path such as `/a/%PLUGIN_ID%/PAGE_NAME`. Any requests sent to /a/%PLUGIN_ID%, for example, `/a/myorgid-simple-app/`, are routed to the root page of the app plugin. The root page is a React component that returns the content for a given route.
```

Let's add a new page to the navigation menu:

1. Modify `plugin.json` to add a new page.

   ```json title="src/plugin.json"
   // ...
   "includes": [
       // ...
       {
           "type": "page",
           "name": "New Page",
           "path": "/a/%PLUGIN_ID%/new-page",
           "roles": "Admin",
           "addToNav": true,
           "defaultNav": false
       }
   ]
   ```

> 1. Save the `src/plugin.json` file

Reviewer 3: ```suggestion
1. Save the `src/plugin.json` file.
```

1. Restart your Grafana instance.

:::note

After saving the `plugin.json` file, you need to restart your Grafana instance to see the new page in the navigation menu.

:::

The new page appears in the navigation menu. You can now edit the React router in `src/components/App/App.tsx` and point a custom component to it.

1. Create a new file called `src/pages/NewPage.tsx` and add the following code:

   ```tsx title="src/pages/NewPage.tsx"
   import React from 'react';
   import { PluginPage } from '@grafana/runtime';

   export function NewPage() {
     return <PluginPage>New Page</PluginPage>;
   }
   ```

1. Modify the routes in `src/components/App/App.tsx` to recognize the new page:

   ```tsx title="src/components/App/App.tsx"
   {
     /* .... */
   }
   <Route path="new-page" element={<NewPage />} />;
   ```

1. Save the file.
1. Reload Grafana to see the new page in place.

You don't need to register all your pages inside `includes` in your `plugin.json`. Register only pages that you want to add to the navigation menu.

:::note

You can only have one level of pages in the navigation menu. Sub-menu items are not supported.

:::

## Configuration page

> You can add a configuration page to your App plugin where users can configure any settings your plugin might need. Your plugin should already have an example configuration page with its source code located in `src/components/AppConfig/AppConfig.tsx`.

Reviewer 3: ```suggestion
You can add a configuration page to your app plugin where users can configure any settings your plugin needs. Your plugin should already have an example configuration page with its source code located in `src/components/AppConfig/AppConfig.tsx`.
```

> ### Saving user settings

Reviewer 3: ```suggestion
### Save user settings
```

To store user settings, send a POST request to `/api/plugins/%PLUGIN_ID%/settings` with the `jsonData` and `secureJsonData` as data.

```ts
export const updatePluginSettings = async (pluginId: string, data: Partial<PluginMeta>) => {
  const response = await getBackendSrv().fetch({
    url: `/api/plugins/${pluginId}/settings`,
    method: 'POST',
    data, // data: { jsonData: { ... }, secureJsonData: { ... } }
  });

  return lastValueFrom(response);
};
```

### Retrieve user settings

The user settings are part of the plugin `meta`. You can retrieve them inside a React component by using the `usePluginContext` hook. For example:

```tsx
import React from 'react';
import usePluginContext from '@grafana/data';

function MyComponent() {
  const context = usePluginContext();
  // user settings
  const jsonData = context.meta.jsonData;
}
```

## Fetch data from frontend code using the data proxy

> If you want to fetch data from your app frontend code (e.g. from a thid party API) without [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) issues or using authenticated requests, then you want to use the [data proxy](../create-a-plugin/extend-a-plugin/fetch-data-from-frontend.md).

Reviewer 3: ```suggestion
If you want to fetch data from your app frontend code (for example, from a thid party API) without [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) issues or using authenticated requests, then you want to use the [data proxy](../create-a-plugin/extend-a-plugin/fetch-data-from-frontend.md).
```

> ## Adding nested plugins inside your app

Reviewer 3: ```suggestion
## Add nested plugins inside your app
```
